### PR TITLE
Use nat_or_var in grammar where negative values don't make sense

### PR DESCRIPTION
**Pull request by jfehrle**
Changes `int_or_var` to `nat_or_var` in the grammar for the following tactics:

- constructor, econstructor, do, hresovle_core, hget_evar - already gave errors for negative values
- {e}auto, info_{e}auto, [ fs | bfs | typeclasses ] eauto - no error but negative max depth is silly
- debug_{e}auto - negatives show unimportant differences in output
- psatz, psatz_ [ Z | R | Q ]

No deprecation on this one, seems a corner case that users are unlikely to use since it doesn't make sense.

Changelog included.

### CODE REVIEW

**Comment by herbelin** [comment]
You probably need at least a `register_grammar wit_nat_or_var (nat_or_var)` in `pltac.ml`.

I suspect that more is needed and if so I'll have a look tomorrow (GMT+1).

**Comment by jfehrle** [comment]
The CI failure doesn't look like a syntax error.  What do you think?

**Comment by Zimmi48** [comment]
> The CI failure doesn't look like a syntax error. What do you think?

Yeah, it's a new failure which has been signaled to metacoq maintainers.

**Comment by Zimmi48** [comment]
It would be very convenient to get this in 8.13 because otherwise the documentation changes on top of this (in particular #13343) will become much harder to backport. If this change doesn't get in 8.13, then a solution could be to delay this PR until #13343 is merged to avoid a rebase that would trigger backporting issues.

@gares WDYT?

**Review by Zimmi48 — APPROVED** [review]
Up to my concern expressed above, this LGTM.

**Comment by Zimmi48** [comment]
I'm tentatively marking this as 8.13+beta1 candidate.

**Comment by gares** [comment]
please merge

**Comment by herbelin** [comment]
@Zimmi48: are you rebasing it?

**Comment by Zimmi48** [comment]
OK, I'll have a look.

**Comment by Zimmi48** [comment]
Rebase done.

@jfehrle I couldn't run `dune @check-gram` but this is because it's already failing on `master`.

Incidentally, this PR fixes the following (unreported AFAICT) anomaly:

```
Goal True.
constructor -1.
(* Anomaly "Uncaught exception Invalid_argument("index out of bounds")." Please report at http://coq.inria.fr/bugs/. *)
```

**Comment by jfehrle** [comment]
> I couldn't run `dune @check-gram` but this is because it's already failing on `master`.

No big deal, we'll pick up any relevant changes soon enough in other PRs.  Later on we can be more fastidious.

> Incidentally, this PR fixes the following (unreported AFAICT) anomaly:

It's probably still there if you pass a variable with a negative value.  If you give me an example for that case, I'll file an issue.   Clearly this is not a variable defined by the `Variable` command.  By the way, somewhere the doc should explain the different types of variables (and namespaces, too) and which type applies to nat-or_var`, etc..

**Comment by Zimmi48** [comment]
Very good point. So that would be:

```coq
Goal True.
let n := -1 in constructor n.
(* Anomaly "Uncaught exception Invalid_argument("index out of bounds")." Please report at http://coq.inria.fr/bugs/. *)
```

The `var` in `nat_or_var` and `int_or_var` are Ltac variables.

**Comment by Zimmi48** [comment]
@herbelin Have you forgotten about this PR? I thought you were asking me to take care of the rebase so that you could merge it without delay.

**Comment by herbelin** [comment]
@coqbot: merge now

**Comment by Zimmi48** [comment]
Shouldn't the `int_or_var` argument of `fail` / `gfail` have been replaced by `nat_or_var` too?

**Comment by jfehrle** [comment]
Yeah, I guess so. See #13469.

**Comment by LasseBlaauwbroek** [comment]
`wit_nat_or_var` is not registered with the printer. See here for the corresponding registration of `wit_int_or_var`: https://github.com/coq/coq/blob/9992bb15d4f90fc0cf3aa2854beb209bc5effac6/plugins/ltac/pptactic.ml#L1319

**Comment by LasseBlaauwbroek** [comment]
In addition: These aliases are causing code duplication for me. As far as I can tell there is no code that actually enforces/checks that these numbers are positive. If that is so, is there a reason why `wit_nat_or_var` is not just an alias of `wit_int_or_var`?

**Comment by LasseBlaauwbroek** [comment]
(Another addition: If that sounded annoyed, it was not meant as such. All the improvements to the manual are making my life much easier!)

**Comment by jfehrle** [comment]
First, you're better off creating new issues rather than adding comments on the end of closed PRs.  Your issues may not noticed if they're buried there.  You can link the issue to the closed PR.

> `wit_nat_or_var` is not registered with the printer. 

An oversight.  Could you provide a small test case showing the problem?  (`Ltac x := do 3 idtac "x".  Print x.` uses `nat_or_var` yet it prints correctly.)

> As far as I can tell there is no code that actually enforces/checks that these numbers are positive. If that is so, is there a reason why `wit_nat_or_var` is not just an alias of `wit_int_or_var`?

That is correct, this was primarily a syntax/doc fix.  Each tactic needs to check that the value coming from a variable is non-negative when the tactic is executed, well after parse time.  I think the `wit_` symbols have to be registered separately so that you can specify `nat_or_var` in `Tactic Notation` and therefore you need do the other registrations.  Just having using an OCaml statement `let wit_nat_or_var = wit_int_or_var` isn't enough.  Also simpler for me (less thinking) to make them independent.

> In addition: These aliases are causing code duplication for me.

Perhaps you could alias some things in your code?

> (Another addition: If that sounded annoyed, it was not meant as such.)

No worries.  But I hope you're not overworking yourself--wasn't your last message sent between 3 and 4 AM?  Sleep, exercise and some recreation are all essential!  As is work.

**Comment by LasseBlaauwbroek** [comment]
> First, you're better off creating new issues rather than adding comments on the end of closed PRs.

Normally I would, but these PR's are so recent that I thought an issue would just be noise.

> Could you provide a small test case showing the problem?
```
Ltac test := constructor 1.
Print test.
```
> Just having using an OCaml statement let wit_nat_or_var = wit_int_or_var isn't enough.

If that is so then I'll have to accept my code duplication. But note that there are already several (non-deprecated) aliases in the code, which makes me think that an alias could be sufficient (if not, these aliases should probably be deprecated):
https://github.com/coq/coq/blob/9992bb15d4f90fc0cf3aa2854beb209bc5effac6/interp/stdarg.ml#L49

> wasn't your last message sent between 3 and 4 AM?

Yes -- sleeping issues :-)

**Comment by jfehrle** [comment]
> But note that there are already several (non-deprecated) aliases in the code, which makes me think that an alias could be sufficient (if not, these aliases should probably be deprecated)

Generally we only bother to deprecate user-visible changes, not OCaml code, though I did recently deprecate some grammar nonterminals.

**Pull request merged**
